### PR TITLE
BUG: Try to sort result of Index.union rather than guessing sortability

### LINKED ISSUE #17376: Union of indexes of tuples is not sorted
#### Code Sample, a copy-pastable example if possible

```python
aidx1 = np.array([(1, 'A'), (2, 'A'), (1, 'B'), (2, 'B')],
                 dtype=[('num', int), ('let', 'a1')])
aidx2 = np.array([(1, 'A'), (2, 'A'), (1, 'B'),
                  (2, 'B'), (1, 'C'), (2, 'C')],
                 dtype=[('num', int), ('let', 'a1')])

idx1 = pd.Index(aidx1)
idx2 = pd.Index(aidx2)

idx1.union(idx2)
```

Result: ``Index([(1, b'A'), (2, b'A'), (1, b'B'), (2, b'B'), (1, b'C'), (2, b'C')], dtype='object')``

This was already known (the example above is copied from [the tests](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/indexes/test_base.py#L1332)), but I don't see it reported anywhere.

#### Problem description

The ``Index.union`` docs states the result is sorted if possible - and it is possible with the above.

#### Expected Output

``` python
In [4]: idx1.union(idx2).sort_values()
Out[4]: Index([(1, b'A'), (1, b'B'), (1, b'C'), (2, b'A'), (2, b'B'), (2, b'C')], dtype='object')
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 0d676a3ccf1d7aa986416a7488b941496f936d98
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-3-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.21.0.dev+405.g0d676a3cc
pytest: 3.0.6
pip: 9.0.1
setuptools: None
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 5.1.0.dev
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.2
openpyxl: None
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.6
lxml: None
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #17376
- [x] tests added / passed
- [x] passes `git diff master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The lines of code I removed seemed written with the intention of guaranteeing the same result under Python 2 and 3, but the docs say "Form the union of two Index objects and sorts _if possible._": so if it is possible in Python 2 we should do it, regardless of whether it is in Python 3. After all, this is how ``.sort_values()`` works. And more practically, trying to guess whether the result of an union can be sorted is hard.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -2195,35 +2195,21 @@ def union(self, other):
...
             else:
                 result = self._values
 
-                try:
-                    result = np.sort(result)
-                except TypeError as e:
-                    warnings.warn("%s, sort order is undefined for "
-                                  "incomparable objects" % e, RuntimeWarning,
-                                  stacklevel=3)
+            try:
```
Comment (jreback) [inline pandas/core/indexes/base.py]: you probably can simply use ``core.sorting.safe_sort`` here instead and just remove the warning entirely (as it fails in many less cases)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (toobaz) [reply pandas/core/indexes/base.py]: Done, but I left the warning as it's still raised if not even ``safe_sort`` can sort them.

(I assume the failure is a problem with AppVeyor)

**Comment by toobaz** [comment]
Any idea about what's happening with AppVeyor? It fails reliably on this PR with 2.7, but the trace doesn't point at any specific test.

**Comment by gfyoung** [comment]
@toobaz : Nope, not a clue :cry: .  There's no way your changes AFAICT could have caused internal errors on the build.  I restarted AppVeyor to double check.

**Comment by toobaz** [comment]
> I restarted AppVeyor to double check.

Failed again... shall I try closing this and opening a new PR?

**Comment by jreback** [comment]
> Failed again... shall I try closing this and opening a new PR?

NO. that just creates more work. let me look.

**Comment by jreback** [comment]
ok this is a legitimate catch. pytest-dist is segfaulting, hiding the traceback. So a couple of options to debug.

setup a vm on 3.6/windows is easiest / best.

you can temporarily change the options in the test runner (in ci/scripts_multiple.py) to something like:
    ``pytest -v -r xX -m "not single" --junitxml=/tmp/multiple.xml $TEST_ARGS pandas``

which disables the multi-processing and shows each test (so you can see where its failing).

**Comment by jreback** [comment]
it *looks* like its either segfaulting or in a recursive loop somewhere.

**Comment by toobaz** [comment]
> setup a vm on 3.6/windows is easiest / best.

Do you mean Python 3.6?! (tests fail only on 2.7)

**Comment by gfyoung** [comment]
@toobaz  : I think @jreback probably meant 2.7 :smile:

**Comment by jreback** [comment]
@toobaz this is related to #17839, could simply recast this to add the ``sort=`` kwarg (I think for back compat we have to use ``=False``, but maybe be able to change that, not sure).

**Comment by toobaz** [comment]
> @toobaz this is related to #17839, could simply recast this to add the sort= kwarg (I think for back compat we have to use =False, but maybe be able to change that, not sure).

Certainly related, but I'm confused: are you suggesting the two methods should share more code? Or just that ``Index.union`` too should have a ``sort=`` argument? In this case I think that for backward compatibility it should default to ``True``...

**Comment by jreback** [comment]
> Certainly related, but I'm confused: are you suggesting the two methods should share more code? Or just that Index.union too should have a sort= argument? In this case I think that for backward compatibility it should default to True...

yes I would add a ``sort=True`` kwarg.

**Comment by jreback** [comment]
can you rebase. I think we do want to add this kw.

**Comment by toobaz** [comment]
Unfortunately the problem on Windows + Python 2.7 persists. I haven't found the time to set up a virtual machine to debug.

**Comment by jreback** [comment]
@toobaz if you can rebase would be great. I would also be ok with simply adding the keyword w/o changing *much* code as a first step.

**Comment by jreback** [comment]
closing as stale. would like to generaly add the ``sort=`` kwarg though.

**Comment by gfyoung** [comment]
@jreback : Let me take a crack at reviving this.

**Comment by toobaz** [comment]
@gfyoung great! If you want a rebase, feel free to ask.

**Comment by gfyoung** [comment]
There's a lint error on Travis, but I want to check the other builds before pushing the fix.

**Comment by gfyoung** [comment]
@toobaz : That windows error looks puzzling.  Could be something with the build itself, as I can't see that "internal error" occurring locally at least.

**Comment by TomAugspurger** [comment]
I don't think it conflicts much with this PR, but FYI I have a branch started adding `sort` to `union`. https://github.com/pandas-dev/pandas/compare/master...TomAugspurger:index-sort

I think it works for every index except `RangeIndex`, since some `RangeIndex.union(RangeIndex)` ops that return a RangeIndex with `sort=True` won't return a `RangeIndex` with `sort=False`.

**Comment by gfyoung** [comment]
@TomAugspurger : Good to know.  Thanks!

**Comment by gfyoung** [comment]
@toobaz : also not sure why you're running up against some random `clip` errors on Travis.  I can't reproduce those locally.

**Comment by TomAugspurger** [comment]
I'm guessing it's a change to NumPy dev:
https://github.com/pandas-dev/pandas/issues/19976 (haven't tried locally
though)

**Comment by gfyoung** [comment]
> I'm guessing it's a change to NumPy dev:

@TomAugspurger : That seems reasonable.  I don't see any recent PR's for `clip`.

**Comment by gfyoung** [comment]
@toobaz : Despite the massive typo in my commit message, the latest commit is a dummy one to determine if that last AppVeyor build is just busted, or there is potentially an issue in the code.  I will undo that commit of course to restore the changes back in the PR.

**Comment by toobaz** [comment]
> @toobaz : Despite the massive typo in my commit message, the latest commit is a dummy one to determine if that last AppVeyor build is just busted, or there is potentially an issue in the code

Fingers crossed... but I had triggered that behavior consistently (on multiple rebases).

**Comment by gfyoung** [comment]
@toobaz : Not looking good here...the build passed just fine when I did all of your changes.

**Comment by gfyoung** [comment]
So adding more verbosity to **pytest** appears to have given a hint as to the problematic tests:

https://ci.appveyor.com/project/pandas-dev/pandas/build/1.0.11213/job/36lok03ygfrk2g86#L5503

**Comment by gfyoung** [comment]
Hmmm...so this issue might be known with pytest (or its plugins) already:

https://github.com/pytest-dev/pytest/issues/2223

Still not sure how @toobaz is triggering that though.

**Comment by toobaz** [comment]
Maybe we could ask whether there is a way to use this [catchsegv](http://www.unix.com/man-page/linux/1/catchsegv/) within a pytest run?

**Comment by gfyoung** [comment]
> Maybe we could ask whether there is a way to use this catchsegv within a pytest run?

@toobaz : Well, this is a Windows machine, not UNIX 😢 

I notice that we are using **pytest-forked**, which AFAICT, isn't well-supported on Windows:

https://github.com/pytest-dev/pytest-forked

I was about to file an issue against **pytest-xdist**, but I wonder if this might be the problem child.

**Comment by toobaz** [comment]
> @toobaz : Well, this is a Windows machine, not UNIX cry

right 🙄

> I was about to file an issue against pytest-xdist, but I wonder if this might be the problem child.

What makes you think this is a bug in pytest, not a segfault somewhere in pandas? (or do you think it's both?)

**Comment by gfyoung** [comment]
> What makes you think this is a bug in pytest, not a segfault somewhere in pandas? (or do you think it's both?)

I'm highly skeptical that your changes are causing a segfault.  That would be pretty impressive actually given that you're only touching high-level Python 😄

**Comment by gfyoung** [comment]
So removing xdist + rebasing somehow resulted in test failures on the first machine?

Going to leave this PR alone for now until `master` is patched,

**Comment by toobaz** [comment]
> I'm highly skeptical that your changes are causing a segfault. That would be pretty impressive actually given that you're only touching high-level Python

Sure, that would require that my changes are triggering a bug somewhere in lower-level code... but I consider this more probable than a bug in pydist which affects precisely that test, in precisely this PR.

Then again, I'm definitely not an expert.

**Comment by gfyoung** [comment]
I suppose, but if it's a bug in your code, I haven't been able to reproduce it locally.

**Comment by toobaz** [comment]
> I suppose, but if it's a bug in your code, I haven't been able to reproduce it locally.

Aha! I had missed that you were testing on Windows. Good to know.

File: appveyor.yml [hunk 2]
```diff
@@ -73,7 +73,7 @@ install:
   - cmd: conda info -a
 
   # create our env
-  - cmd: conda create -n pandas python=%PYTHON_VERSION% cython pytest>=3.1.0 pytest-xdist
+  - cmd: conda create -n pandas python=%PYTHON_VERSION% cython pytest>=3.1.0
```
Comment (jreback) [inline appveyor.yml]: so does xdist cause the issues with this?

**Comment by gfyoung** [comment]
@toobaz @jreback : So good news (I guess?), I was able to replicate the AppVeyor failure locally.

**Comment by gfyoung** [comment]
@toobaz : It appears that you loosened the restrictions too much, at least for Windows:

* The test that is breaking AppVeyor is **TestAlignment.test_basic_frame_alignment**
* The condition where **len(indexer) > 0** is the problem.  For some reason, I need to put back the comparison of `lvals[0]` and `other_diff[0]` to remove the break OR only trying sorting when in the `else` branch (i.e. when `len(indexer) = 0`).

**Comment by toobaz** [comment]
@gfyoung thanks for your analysis!

Does the problem arise with any specific combination of index types? (from ``args = product(...``)

**Comment by gfyoung** [comment]
> Does the problem arise with any specific combination of index types?

I can't tell unfortunately.  It seems to have crashed with multiple types.

**Comment by jreback** [comment]
@toobaz whats' the status on this?

**Comment by jreback** [comment]
closing as stale, if you want to continue working, pls ping and we can re-open. you will need to merge master.

**Pull request closed without merging**